src/operations/blas2/gemv.hpp

```diff
@@ -80,6 +80,244 @@ SYCL_BLAS_INLINE void AddSetColumns<rhs_t>::adjust_access_displacement() {
+  const index_t nc_dim_index = local_id + nc_group_id * local_range;
+
+  // Stores the sum for the partial dot product computation
+  value_t sum = 0;
+
+  // In the non transposed case
+  if (!is_transposed) {
+
+    // Make sure nc_dim_index is within bounds
+    // This closes threads that would be calculating beyond the final row
+    if (nc_dim_index >= nc_dim) {
+      return 0;
+    }
+
+    // Calculate the matrix index
+    index_t mat_index = nc_dim_index + c_group_id * local_range * lda;
+    
+    const index_t last_c_dim_id = std::min(c_dim - c_group_id * local_range, local_range);
+
+    // Computes the partial dot product for a row
+    for (index_t c_dim_id = 0; c_dim_id < last_c_dim_id; ++c_dim_id) {
+      sum += ProductOperator::eval(matrix_.template eval<true>(mat_index), local_mem[c_dim_id]);
+      mat_index += lda;
+    }
+
+    const index_t out_index = nc_dim_index + (c_group_id * nc_dim);
+    return lhs_.eval(out_index) = sum;
+  }
+  else { // In the transposed case 
+
+    constexpr int cache_capacity = cache_line_size / sizeof(value_t);
+
+    const int tile_c_loops = local_range / cache_capacity;
+
+    // Pointer to scratch space used for loading input matrix and transposing it on-the-fly
+    value_t* matrix_scratch = local_mem.localAcc.get_pointer() + local_range;
+
+    for (index_t c_tile_id = 0; c_tile_id < tile_c_loops; ++c_tile_id) {
+
+      // Extract a matrix block from global memory
+      extract_input_block(matrix_scratch, ndItem, c_tile_id);
+
+      // Ensure memory synchronization within work group
+      ndItem.barrier(cl::sycl::access::fence_space::local_space);
+
+      index_t mat_index = local_id;
+
+      #pragma unroll
+      for (index_t c_dim_id = 0; c_dim_id < cache_capacity; ++c_dim_id) {
+
+        const index_t vec_index = c_tile_id * cache_capacity + c_dim_id;
+
+        sum += ProductOperator::eval(matrix_scratch[mat_index], local_mem[vec_index]);
+
+        mat_index += local_range + 1; // Adding one as bank offset
+      }
+
+      // Ensure memory synchronization within work group
+      ndItem.barrier(cl::sycl::access::fence_space::local_space);
+    }
+
+    const index_t out_index = nc_dim_index + (c_group_id * nc_dim);
+    
+    if (nc_dim_index < nc_dim) {
+      lhs_.eval(out_index) = sum;
+    }
+    return sum;
+  } 
+}
+
+template <typename lhs_t, typename matrix_t, typename vector_t, bool is_transposed, int cache_line_size, int work_per_thread>
+template <typename ScratchPointerType>
+SYCL_BLAS_INLINE
+void
+Gemv<lhs_t, matrix_t, vector_t, is_transposed, cache_line_size, work_per_thread>::extract_input_block(ScratchPointerType matrix_scratch, cl::sycl::nd_item<1> id, index_t c_tile_id) {
+
+  constexpr int cache_capacity = cache_line_size / sizeof(value_t);
+
+  const index_t local_id = id.get_local_id(0);
+  const index_t group_id = id.get_group(0);
+  const index_t local_range = id.get_local_range(0);
+  const index_t lda = matrix_.getSizeL();
```

Comment:
Seems like these could be passed in as parameters as the values are already obtained inside eval() where this function is called.

Reply:
Done for `local_id`, `group_id`, and `local_range`. Don't think it's necessary for `lda` since `matrix_` is a member and `getSizeL()` is inlined ?

Reply:
That is true, though I would probably replace it anyway just to avoid the duplication, I'll leave it up to you though!

Reply:
I've made it a parameter. Though I'm still unsure of which option would result in more performant code. Either way the difference should be miniscule.